### PR TITLE
JIT: Catch "cns".Length also in Morph

### PR DESCRIPTION
**Pull request by EgorBo**
I noticed that the existing `"const".Length` optimization doesn't handle this case well: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAZgAJiAmcgYQFgAoAbyfPcoEYAGcgWQCeAZQxRyAXgB85AEQALGABtFEGQG42HTewCWAOwzkAKjFwYAFAEoJ0wSKgA6ADIw9AcwxyNjAL5A=

Jit-diff is small:
```
PMI CodeSize Diffs for System.Private.CoreLib.dll, framework assemblies for  default jit

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 52988913
Total bytes of diff: 52988811
Total bytes of delta: -102 (-0.00% of base)
    diff is an improvement.


Top file improvements (bytes):
         -27 : System.Private.CoreLib.dasm (-0.00% of base)
         -22 : System.Resources.Extensions.dasm (-0.06% of base)
         -22 : System.Resources.Writer.dasm (-0.27% of base)
         -13 : System.IO.Packaging.dasm (-0.01% of base)
         -11 : System.Reflection.Metadata.dasm (-0.00% of base)
          -7 : System.Private.Xml.dasm (-0.00% of base)

6 total files with Code Size differences (6 improved, 0 regressed), 265 unchanged.

Top method improvements (bytes):
         -27 (-6.21% of base) : System.Private.CoreLib.dasm - UTF7Encoding:MakeTables():this
         -22 (-0.91% of base) : System.Resources.Extensions.dasm - PreserializedResourceWriter:Generate():this
         -22 (-0.90% of base) : System.Resources.Writer.dasm - ResourceWriter:Generate():this
         -13 (-3.08% of base) : System.IO.Packaging.dasm - ContentTypeHelper:OpenContentTypeStream(ReadOnlyCollection`1):Stream:this
         -11 (-7.14% of base) : System.Reflection.Metadata.dasm - PortablePdbBuilder:.ctor(MetadataBuilder,ImmutableArray`1,MethodDefinitionHandle,Func`2):this
          -7 (-6.25% of base) : System.Private.Xml.dasm - Base64Decoder:ConstructMapBase64():ref

Top method improvements (percentages):
         -11 (-7.14% of base) : System.Reflection.Metadata.dasm - PortablePdbBuilder:.ctor(MetadataBuilder,ImmutableArray`1,MethodDefinitionHandle,Func`2):this
          -7 (-6.25% of base) : System.Private.Xml.dasm - Base64Decoder:ConstructMapBase64():ref
         -27 (-6.21% of base) : System.Private.CoreLib.dasm - UTF7Encoding:MakeTables():this
         -13 (-3.08% of base) : System.IO.Packaging.dasm - ContentTypeHelper:OpenContentTypeStream(ReadOnlyCollection`1):Stream:this
         -22 (-0.91% of base) : System.Resources.Extensions.dasm - PreserializedResourceWriter:Generate():this
         -22 (-0.90% of base) : System.Resources.Writer.dasm - ResourceWriter:Generate():this
```

### CODE REVIEW

**Comment by SingleAccretion** [comment]
Suggestion: maybe introduce something like `gtFoldCnsStrLength(GenTree* tree)` that would transform `tree` to a constant, and use it here and in the importer? This would be a nice place to have the `JITDUMP` as well.

**Review by SingleAccretion — APPROVED** [review]

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -18417,6 +18407,28 @@ void Compiler::impWalkSpillCliqueFromPred(BasicBlock* block, SpillCliqueWalker*
     assert(impSpillCliqueGetMember(SpillCliquePred, block) != 0);
 }
 
+GenTreeIntCon* Compiler::impStringLiteralLength(GenTreeStrCon* node)
```
Comment (BruceForstall) [inline src/coreclr/jit/importer.cpp:18410]: Please add a function comment

File: src/coreclr/jit/morph.cpp [hunk 2]
```diff
@@ -12500,6 +12500,20 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
 #endif // !TARGET_64BIT
             break;
 
+        case GT_ARR_LENGTH:
+            if (op1->OperIs(GT_CNS_STR))
+            {
+                // Optimize `ldstr + String::get_Length()` to CNS_INT
+                // e.g. "Hello".Length => 5
+                GenTreeIntCon* iconNode = impStringLiteralLength(op1->AsStrCon());
```
Comment (BruceForstall) [inline src/coreclr/jit/morph.cpp:12508]: Seems odd to call a function from morph that is "impX", although I guess it wouldn't be the first case.

Comment (EgorBo) [reply src/coreclr/jit/morph.cpp:12508]: Yeah there are already quite a few:
```
impGetRefAnyClass
impInlineRoot
impTailCallRetTypeCompatible
impFixupStructReturnType
impCheckForNullPointer
impIsAddressInLocal
impReadyToRunHelperToTree
```

Comment (AndyAyersMS) [reply src/coreclr/jit/morph.cpp:12508]: Even so, it seems like it belongs over in gentree.cpp.

File: src/coreclr/jit/importer.cpp [hunk 3]
```diff
@@ -18417,6 +18407,38 @@ void Compiler::impWalkSpillCliqueFromPred(BasicBlock* block, SpillCliqueWalker*
...
+    int             length = -1;
+    const char16_t* str    = info.compCompHnd->getStringLiteral(node->gtScpHnd, node->gtSconCPX, &length);
+    if (length >= 0)
+    {
+        GenTreeIntCon* iconNode = gtNewIconNode(length);
+
+        // str can be NULL for dynamic context
+        if (str != nullptr)
+        {
+            JITDUMP("Optimizing '\"%ws\".Length' to just '%d'\n", str, length)
```
Comment (AndyAyersMS) [inline src/coreclr/jit/importer.cpp:18431]: NIT: this doesn't actually do any optimization, so maybe change this to
```suggestion
            JITDUMP("String '\"%ws\".Length' is '%d'\n", str, length)
```

**Comment by EgorBo** [comment]
@dotnet/jit-contrib PTAL

**Review by AndyAyersMS — APPROVED** [review]

**Pull request merged**